### PR TITLE
VSTS CI/CD support

### PR DESCRIPTION
**Pull request by bartr**
additional unit tests
support for VSTS CI/CD
bug fixes
changed folder structure go make VSTS work correctly and make go tools work better

### CODE REVIEW

File: docker/debug/runapp.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+cd /root
+
+service ssh start
```
Comment (day-jeff) [inline docker/debug/runapp.sh]: I know this is "only" a sample app, but it still seems like the developer should take an explicit step (e.g. pass a parameter) to enable ssh. Starting it by default gives me the heebie jeeebies.

File: docker/debug/runapp.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+mkdir -p /home/LogFiles
+
+file=restart.txt
+echo "restart" > $file
+
+echo "\nStarting server ($WEBSITE_ROLE_INSTANCE_ID) ..."
```
Comment (day-jeff) [inline docker/debug/runapp.sh]: Where is $WEBSITE_ROLE_INSTANCE_ID set? What is it?

File: docker/debug/runapp.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+cd /root
+
+service ssh start
+
+mkdir -p /home/LogFiles
```
Comment (day-jeff) [inline docker/debug/runapp.sh]: As mentioned elsewhere, I recommend we make the log path a parameter.

File: docker/debug/runapp.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+echo "restart" > $file
+
+echo "\nStarting server ($WEBSITE_ROLE_INSTANCE_ID) ..."
+
+# git the latest code
+cd /root/m4
+git pull
+cd /root
+
+while [ -f "$file" ]
```
Comment (day-jeff) [inline docker/debug/runapp.sh]: Not clear to me why you're looping and restarting the app. Does it kill itself do to errors? Would be good to add comment why this is needed.

File: sendmessage/main.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
+package main
```
Comment (day-jeff) [inline sendmessage/main.go]: Aha...now I see why my comments on the previous PR were "eaten"--you included these files in this PR. I don't think that is proper practice. I think the better way to do it is commit any updates to unmerged files back to the pending PR (it just works!). This way reviewer comments aren't lost and there aren't conflicting versions pending.

File: sendmessage/main.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	// read the response
+	_, err = ioutil.ReadAll(resp.Body)
+
+	if err != nil {
```
Comment (day-jeff) [inline sendmessage/main.go]: Wondering if it's overkill to error check .ReadAll if .Do has just succeeded.

File: sendmessage/main.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+	// execute and get a response
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		log.Println(err)
+		return
+	}
```
Comment (day-jeff) [inline sendmessage/main.go]: Opportunity to be more concise with error checking:

	if resp, err := client.Do(req); err != nil {
	    log.Println(err)
	    return
	}

This is my style preference as it feels "cleaner" to me--curious what you/the CSE style guide have to say.

File: sendmessage/main.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	// read the response
+	_, err = ioutil.ReadAll(resp.Body)
+
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	// close the request
+	resp.Body.Close()
```
Comment (day-jeff) [inline sendmessage/main.go]: Probably better to put this write after the .Do() error check using a defer statement.

File: src/m4/sampleapp/main_test.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	env.EventTime = time.Now().UTC().Format("2006-01-02T15:04:05Z")
+
+	p := person{FirstName: "John", LastName: "Doe"}
+
+	var err error
+	env.Data, err = json.Marshal(&p)
+
+	if err != nil {
+		t.Error(err)
+	}
```
Comment (day-jeff) [inline src/m4/sampleapp/main_test.go]: Overkill to error check .March from strongly typed struct? I don't think there will ever be an error in that case.

File: docker/debug/runapp.sh (hunk 1 above)
Comment (bartr) [reply docker/debug/runapp.sh]: this is the "debug" dockerfile and this allows you to debug the code running in App Services. You can only get to the SSH support in app services through the Azure portal and have to have access to the subscription.

the release docker file is locked down

File: docker/debug/runapp.sh (hunk 2 above)
Comment (bartr) [reply docker/debug/runapp.sh]: It's set by App Services - you can have multiple instances of an app running in App Services (scale out) and each one gets assigned a unique number.

File: docker/debug/runapp.sh (hunk 3 above)
Comment (bartr) [reply docker/debug/runapp.sh]: I'll remove that line from the script

File: docker/debug/runapp.sh (hunk 4 above)
Comment (bartr) [reply docker/debug/runapp.sh]: this is the debug version. The use case is a developer is debugging the app running in App Services using SSH. You want to make a change to debug an issue, so if you run ps -A and kill the pid associated with sampleapp, it restarts automatically and immediately. This is way faster than going through a full deploy.

This is NOT in the release docker file. It just runs the app directly.

that's also why we use the full golang base and install dev tools in the debug build - so you can actually debug in App Services as some things (like getting a message from Event Grid) are hard to simulate on your dev box.

File: sendmessage/main.go (hunk 5 above)
Comment (bartr) [reply sendmessage/main.go]: this branch is based off the eventgrid branch, so all changes are already included. This is additional functionality beyond what was in the eg branch.

Not sure if it's "proper" or not, but the eg branch was already huge due to the refactoring, so I didn't want to make it even larger.

File: sendmessage/main.go (hunk 6 above)
Comment (bartr) [reply sendmessage/main.go]: ReadAll could fail - out of memory, closed socket, etc.

File: src/m4/sampleapp/main_test.go (hunk 9 above)
Comment (bartr) [reply src/m4/sampleapp/main_test.go]: probably, but, generally if a function returns an error, I check it.

File: sendmessage/main.go (hunk 7 above)
Comment (bartr) [reply sendmessage/main.go]: I'm OK either way, but we should be consistent. I see it both ways in Go code.

File: sendmessage/main.go (hunk 5 above)
Comment (day-jeff) [reply sendmessage/main.go]: I think you misunderstand me--I made a bunch of comments on the sample app files in the PR for the event grid branch. When you re-commited the same unmerged files in this branch, those files and my review were effectively deleted from the PR. I then re-made several of my comments in this PR, only to learn that the files had been removed from this PR.

What I'm saying is that best practice is change the files in only one PR until they are merged.

Make sense?

File: sendmessage/main.go (hunk 8 above)
Comment (bartr) [reply sendmessage/main.go]: I just updated and pushed

File: docker/debug/runapp.sh (hunk 4 above)
Comment (day-jeff) [reply docker/debug/runapp.sh]: Hmmm, my earlier reply didn’t post. 

I would expect developers to manually kill and rerun the app while in the container via ssh.

**Comment by day-jeff** [comment]
Merge it! (No “approve” option on the mobile UI)

File: docker/debug/runapp.sh (hunk 4 above)
Comment (bartr) [reply docker/debug/runapp.sh]: if you kill the app, App Services will pull the new docker image and restart the app which will kill your SSH session and you will lose any changes you might have made. Plus it takes a minute or two to restart vs. a few seconds.

One way I used this is to SSH in, change git branches, git pull, kill the pid. I just repeat on my dev branch until I got it working. WAY faster than the full CI/CD pipeline.

You can also nano a small file directly, but not larger files.

It's a pretty compelling debugging environment. Not as good as a remote VSTS debugger, but better than nothing.

**Pull request merged**